### PR TITLE
TLS and certificates

### PR DESCRIPTION
**Pull request by ieperen3039**

### CODE REVIEW

**Comment by rvolosatovs** [comment]
@ieperen3039 are you going to rebase this?

File: certificates/cert.pem [hunk 1]
```diff
@@ -0,0 +1,18 @@
+-----BEGIN CERTIFICATE-----
```
Comment (rvolosatovs) [inline certificates/cert.pem]: remove this

File: certificates/main.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
+// Copyright 2009 The Go Authors. All rights reserved.
```
Comment (rvolosatovs) [inline certificates/main.go]: Is this just copy-pasted from go source?

File: cmd/srrs/main.go [hunk 3]
```diff
@@ -157,15 +159,23 @@ func main() {
 			mux.Handle("/", http.FileServer(http.Dir(*static)))
 		}
 
+		// unsecure server
```
Comment (rvolosatovs) [inline cmd/srrs/main.go]: `insecure`

File: cmd/srrs/main.go [hunk 4]
```diff
@@ -157,15 +159,23 @@ func main() {
...
 		tcpLogger := logger.With(zap.String("listen_addr_tcp", *tcpAddr))
 		tcpSrv := &http.Server{
 			Addr:     *tcpAddr,
-			ErrorLog: zap.NewStdLog(logger),
+			ErrorLog: zap.NewStdLog(tcpLogger),
 			Handler:  mux,
 		}
 
-		// TODO: Create tlsServ
+		// identical to tcpSrv
```
Comment (rvolosatovs) [inline cmd/srrs/main.go]: `// secure server`

File: cmd/srrs/main.go [hunk 5]
```diff
@@ -157,15 +159,23 @@ func main() {
...
+		tlsLogger := logger.With(zap.String("listen_addr_tcp", *tlsAddr))
+		tlsSrv := &http.Server{
+			Addr:     *tlsAddr,
+			ErrorLog: zap.NewStdLog(tlsLogger),
+			Handler:  mux,
+		}
 
+		// unsecure server
```
Comment (rvolosatovs) [inline cmd/srrs/main.go]: you already had this comment on L162

File: cmd/srrs/main.go [hunk 6]
```diff
@@ -176,7 +186,13 @@ func main() {
 
 		tlsErrCh := make(chan error, 1)
 		go func() {
-			// TODO: Start tlsServ
+			cert := filepath.Join("..", "..", "certificates", "cert.pem")
+			key := filepath.Join("..", "..", "certificates", "key.pem")
```
Comment (rvolosatovs) [inline cmd/srrs/main.go]: these 2 things should be cofigurable

File: pkg/trcapi/pool.go [hunk 7]
```diff
@@ -7,6 +7,8 @@ import (
 )
 
 // Pool represents a pool of Conn's.
+// Creating a new Pool allows new connections to be established based on given settings.
+// The Pool manages synchronisation and allows for easy access and closing of connections
```
Comment (rvolosatovs) [inline pkg/trcapi/pool.go]: `.` at the end

File: pkg/trcapi/pool.go [hunk 8]
```diff
@@ -7,6 +7,8 @@ import (
 )
 
 // Pool represents a pool of Conn's.
+// Creating a new Pool allows new connections to be established based on given settings.
```
Comment (rvolosatovs) [inline pkg/trcapi/pool.go]: There are no settings for `Pool`, those are functions

File: pkg/trcapi/conn.go [hunk 9]
```diff
@@ -59,7 +59,8 @@ type Conn struct {
 }
 
 // Connect establishes the SRRS-side connection according to TRC API protocol
-// specification of version ver on w and r.
+// specification of version ver.
+// Messages are written to w and read from r
```
Comment (rvolosatovs) [inline pkg/trcapi/conn.go]: `.` at the end

File: cmd/srrs/main.go (hunk 6 above)
Comment (ieperen3039) [reply cmd/srrs/main.go]: will be in a future version

File: certificates/main.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+	if len(*host) == 0 {
+		log.Fatalf("Missing required --host parameter")
+	}
+
+	var priv interface{}
+	var err error
+	switch *ecdsaCurve {
+	case "":
+		priv, err = rsa.GenerateKey(rand.Reader, *rsaBits)
```
Comment (HaicoDorenbos) [inline certificates/main.go]: Are you sure this won't go wrong? RSA uses other private keys than ECDSA. ECDSA only need 256 bits key size to be secure.

File: certificates/main.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to parse creation date: %s\n", err)
+			os.Exit(1)
+		}
+	}
+
+	notAfter := notBefore.Add(*validFor)
+
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
```
Comment (HaicoDorenbos) [inline certificates/main.go]: rand.Int will return an integer. An integer has size 32 bits, but I think you want to generate a biginteger with size at most 128 bits

File: certificates/key.pem [hunk 12]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
[the remark this answers is not in the document]
Comment (helaan) [reply certificates/key.pem]: It doesn't matter how big it is, if you post the private key, it is leaked anyway. See Superfish

File: .gitignore [hunk 13]
```diff
@@ -3,6 +3,7 @@ release
 
 *.pdf
 *.jpg
+*.pem
```
Comment (helaan) [inline .gitignore]: To get the current pem file out of there, use `git rm certificate/cert.pem`.

File: cmd/srrs/main.go [hunk 14]
```diff
@@ -174,10 +177,29 @@ func main() {
...
+		if cert != "" && key != "" {
+			tlsLogger := logger.With(zap.String("listen_addr_tcp", *tlsAddr))
+			tlsSrv := &http.Server{
+				Addr:     *tlsAddr,
+				ErrorLog: zap.NewStdLog(tlsLogger),
+				Handler:  mux,
+			}
+
+			go func() {
+				tlsLogger.With(zap.String("certificate path", cert), zap.String("key path", key))
```
Comment (rvolosatovs) [inline cmd/srrs/main.go]: this does not do anything. you can simply specify the fields on the line below (i.e. `tlsLogger.Info("...", zap.String(..., ...), zap.String(..., ...))`)

File: cmd/srrs/main.go [hunk 15]
```diff
@@ -174,10 +177,29 @@ func main() {
...
+				tlsLogger.Info("Starting the secure web server...")
+				if err := tlsSrv.ListenAndServeTLS(cert, key); err != nil {
+					tlsErrCh <- errors.Wrap(err, "failed to listen")
+				}
+			}()
+
+		} else {
+			logger.Debug("Certificate not specified; Skipping TLS server")
```
Comment (rvolosatovs) [inline cmd/srrs/main.go]: Maybe make this an `info` level and indicate which is missing(key or cert)

**Pull request merged**